### PR TITLE
[Mobile Payments] Announce Card Present Payments feature on Payment Methods screen

### LINKED ISSUE #7229: [Mobile Payments] Announce Card Present Payments feature on Payment Methods screen
### Description
We will use some of the blank space on the Payment Methods screen to publicise the availability of Card Present Payments.

We'll use a banner to announce the feature, with a CTA to purchase a card reader as the first step towards taking payments. This will use the existing web view approach.

upsell-designs

- [x] Show a banner on the Payment Selection screen
- [x] Display a header – “Accept payments easily”.
- [x] Display text – “Get ready to accept payments with a card reader.”.
- [x] Display an illustration.
- [x] Display an action button – “Purchase Card Reader”.
- [x] Opens Purchase Card Reader Webview flow on tap.
- [x] Navigates to [https://woocommerce.com/products/hardware/](https://woocommerce.com/products/hardware/)[COUNTRY] no matter the Payment Gateway
- [x] Display the banner only for stores in the US and CA.

See pdfdoF-ZW-p2 for further context

### PR DESCRIPTION
**Pull request by toupper**
Closes: #7229 

### Description

Adds a banner for the first iteration of upselling Card Present Payments in the app. This banner is on the Payment Method screen, and is permanently shown: we can look to remove it in future when it is no longer useful, but for now there is plenty of space and it is below all the actions on that screen.

This banner only shows for stores in countries where Card Present Payments is supported.

### Testing instructions

1. Launch the app and go to the Orders tab
2. Tap `+` and select `Simple Payment` or `Create Order
3. Set up the payment/order and go through to the payment method selection screen.
4. If your store is in the US or Canada, observe that the banner is shown. Otherwise, observe that it isn't, and stop.
5. Tap on the CTA
6. Observe that the web view opens and loads the correct product page for your store's country: M2 reader for US, WisePad for Canada
7. Observe that you can tap buy, log in, and get to the checkout.

### Screenshots

https://user-images.githubusercontent.com/2472348/177575409-b98d1fbe-0944-4aac-9aa3-3dde594f8dc6.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by joshheald — APPROVED** [review]
Most of this is @toupper's work, I just finished it off.

@iamgabrielma / @malinajirka I think it'd be good to get a once-over from one or both of you before I merge too, since I did work on it too.

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/FeatureAnnouncementView.swift [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import Foundation
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/FeatureAnnouncementView.swift]: I believe we do not need to import Foundation in this view?

File: WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift [hunk 2]
```diff
@@ -44,6 +44,14 @@ final class PaymentMethodsViewModel: ObservableObject {
         paymentLink != nil
     }
 
+    var showUpsellCardReaderFeatureBanner: Bool {
+        return cardPresentPaymentsConfiguration.isSupportedCountry
+    }
+
+    var purchaseCardReaderUrl: URL {
+        cardPresentPaymentsConfiguration.purchaseCardReaderUrl()
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift]: Perhaps we should add inline comments here? What do you think of something like:

```
/// Defines if the Card Reader upsell banner should be shown based on country eligibility
///
var showUpsellCardReaderFeatureBanner: Bool {
	return cardPresentPaymentsConfiguration.isSupportedCountry
}

/// Returns the URL where the merchant can purchase a card reader based on store country code
///
var purchaseCardReaderUrl: URL {
	cardPresentPaymentsConfiguration.purchaseCardReaderUrl()
}
```

File: WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsView.swift [hunk 3]
```diff
@@ -156,6 +172,12 @@ extension PaymentMethodsView {
         static let markAsPaidTitle = NSLocalizedString("Mark as Paid?", comment: "Alert title when selecting the cash payment method")
         static let markAsPaidButton = NSLocalizedString("Mark as Paid", comment: "Alert button when selecting the cash payment method")
         static let cancelTitle = NSLocalizedString("Cancel", comment: "Title for the button to cancel the payment methods screen")
+        static let featureAnnouncementTitle = NSLocalizedString("Accept payments easily",
+                    comment: "Title for the feature announcement banner intended to upsell card readers")
+        static let featureAnnouncementMessage = NSLocalizedString("Get ready to accept payments with a card reader.",
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsView.swift]: In both of these strings we're using "payments" as plural but in pdfdoF-ZW-p2 all references are to "payment" in singular. I'm not a native speaker so I'm not sure what would be more correct in this case, just wanted to point out the difference 👍

File: WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsView.swift [hunk 4]
```diff
@@ -35,33 +37,44 @@ struct PaymentMethodsView: View {
...
+                        }
                     }
                 }
+                .padding(.horizontal)
+                .background(Color(.listForeground))
+
+                if viewModel.showUpsellCardReaderFeatureBanner {
+                    FeatureAnnouncementView(title: Localization.featureAnnouncementTitle,
+                                            message: Localization.featureAnnouncementMessage,
+                                            buttonTitle: Localization.featureAnnouncementButtonTitle,
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsView.swift]: Should we be more specific with these constant names? For example:

- `featureAnnouncementTitle` -> `purchaseCardReaderAnnouncementTitle`
- `featureAnnouncementMessage` -> `purchaseCardReaderAnnouncementMessage`
- `featureAnnouncementButtonTitle` -> `purchaseCardReaderAnnouncementButtonTitle`

**Review by iamgabrielma — APPROVED** [review]

**Comment by iamgabrielma** [comment]
I commented a few nits, but everything works as expected and looks great 💯

**Comment by pachlava** [comment]
I did a somewhat retroactive testing, and found two minor UI issues, found only on iPhone, in landscape mode (I'm having troubles with real device installable, which I hope will be sorted out today, so these are from iPhone 13 Simulator):

Screenshot 2022-07-07 at 12 47 47

- The view was not scrollable initially, which was fine when there was no banner. Now, when it's there, you can notice that `Choose your payment method` label is placed a bit off.
- More obvious when notch is on the left - the `NEW` bubble and the texts beneath should be X-aligned with the elements above them (as it happens in portrait mode).

Please let me know if I should create an issue for that.

Also, a bit delayed question, since the PR is already reviewed and merged. It's the first time I'm working with WC app on the stage earlier than beta testing, so I wonder what is the process of adding the unit tests? Is this usually done along with the PR or in separate? If along, how the decision of adding/not adding is made? Thanks!

**Comment by joshheald** [comment]
> I did a somewhat retroactive testing, and found two minor UI issues, found only on iPhone, in landscape mode (I'm having troubles with real device installable, which I hope will be sorted out today, so these are from iPhone 13 Simulator):
>
> Please let me know if I should create an issue for that.

Thanks @pachlava! I did make an issue (#7246) and I've also done the fix for it in #7247.

Personally, I like to have an issue for anything that isn't totally trivial to fix, i.e. if it's a 5-minute obvious 1-liner, I'll do it without an issue. In this case, it took me about an hour of fiddling around.

I'm happy to make issues myself when I know which of these it is. The comment you left was perfect, which you can see because I didn't feel the need to add any more detail to the ticket when I made it! 

I'm not sure how others on the team feel about it though, I'm afraid...!


> Also, a bit delayed question, since the PR is already reviewed and merged. It's the first time I'm working with WC app on the stage earlier than beta testing, so I wonder what is the process of adding the unit tests? Is this usually done along with the PR or in separate? If along, how the decision of adding/not adding is made? Thanks!

We aim to unit test our work where it makes sense. In this PR, there wasn't anything which it seemed valuable to unit test, because the logic is trivial, and most of the PR is UI, which we wouldn't unit test.

As it becomes more complex, I'd expect to add unit tests where possible, but we do that in whatever way makes the most sense for the work, there's no strict rules to say that they must be in the feature PR or they must be done separately. 

IMO, unit tests are primarily a development tool, rather than a real test of the functionality, and so it's at the discretion of the developer when, what, and how much to unit test... but again, I'm not sure how widely held that view is.

**Comment by pachlava** [comment]
@joshheald thank you for answering, as always, with such great care (and also, for creating and fixing the issue) ⭐

**Pull request merged**